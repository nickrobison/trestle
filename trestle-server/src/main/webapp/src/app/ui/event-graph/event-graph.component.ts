import {AfterViewInit, Component, ElementRef, Input, OnChanges, SimpleChanges, ViewChild} from "@angular/core";
import {BaseType, select, Selection} from "d3-selection";
import {scaleLinear, scaleOrdinal, scaleTime} from "d3-scale";
import {schemeCategory10} from "d3";
import {axisBottom} from "d3-axis";
import {ID3Margin} from "../common";

export interface IEventElement {
  id: string;
  entity: string;
  temporal: Date;
  bin: number;
  value: any;
  continuing?: boolean;
}

export interface IEventLink {
  source: IEventElement;
  target: IEventElement;
}

export interface IEventData {
  bins: number;
  nodes: IEventElement[];
  links: IEventLink[];
}

@Component({
  selector: 'event-graph',
  templateUrl: './event-graph.component.html',
  styleUrls: ['./event-graph.component.scss']
})
export class EventGraphComponent implements AfterViewInit, OnChanges {
  @ViewChild('graph') public element: ElementRef;
  @ViewChild('tooltip') private tooltipEl: ElementRef;
  @Input() public selectedIndividual: string;
  @Input() public data: IEventData;
  @Input() public graphHeight: number;
  @Input() public graphWidth: number;
  @Input() public minDate: Date;
  @Input() public maxDate: Date;
  @Input() public filterLabel?: (input: IEventElement) => string;
  private htmlElement: HTMLElement;
  private host: Selection<HTMLElement, IEventElement | IEventLink, null, undefined>;
  private svg: Selection<BaseType, IEventElement | IEventLink, null, undefined>;
  private width: number;
  private height: number;
  private margin: ID3Margin;
  private tooltip: any;


  public constructor() {

  }

  public ngAfterViewInit(): void {
    this.htmlElement = this.element.nativeElement;
    this.setupD3();
  }

  public ngOnChanges(changes: SimpleChanges): void {
    const dataChange = changes['data'];
    if (dataChange.currentValue !== undefined
      && (dataChange.currentValue !== dataChange.previousValue)) {
      console.debug('Changed, plotting data', this.data);
      this.plotData();
    }
  }

  private plotData(): void {
    // Nuke everything, because I can't figure out the update pattern
    this.svg.selectAll('*').remove();
    //    Setup the X/Y/Z values
    const entityNames = this.data.nodes.map((d) => d.entity);
    const x = scaleTime()
      .range([120, this.width])
      .domain([this.minDate, this.maxDate]);
    const y = scaleLinear()
      .range([this.height, 0])
      .domain([0, this.data.bins]);
    const z = scaleOrdinal(schemeCategory10)
      .domain(entityNames);

    // Filter down the events, to make sure they don't go off the screen
    this.data.nodes.map((node) => {
      if (node.temporal > this.maxDate) {
        node.continuing = true;
        node.temporal = this.maxDate;
      } else if (node.temporal < this.minDate) {
        node.continuing = true;
        node.temporal = this.minDate;
      }
    });

    // Let's draw a box around the selectHandler selection
    const focusedIndividual = this.svg.selectAll<SVGRectElement, BaseType>('.selected')
      .data(this.data.nodes
          .filter((node) => node.entity === this.selectedIndividual),
        (d: IEventElement) => d.entity);

    // Calculate bin height
    const binHeight = y(1) - y(2);

    focusedIndividual
      .enter()
      .append('rect')
      .attr('class', 'selected')
      .attr('x', 0)
      .attr('y', (d) => y(d.bin + 1) + (binHeight / 2))
      .attr('height', binHeight)
      .attr('width', () => x(this.maxDate) + 120)
      .attr('fill', '#FAFAD2')
      .attr('opacity', '0.7')
      .merge(focusedIndividual);

    focusedIndividual
      .exit()
      .remove();

    //    Add the lines
    const links = this.svg
      .selectAll<SVGLineElement, BaseType>('.link')
      .data(this.data.links, (link: IEventLink) => link.source.id + '_' + link.target.id);

    links
      .enter()
      .append('line')
      .attr('class', 'link')
      .attr('x1', (d) => x(d.source.temporal))
      .attr('y1', (d) => y(d.source.bin) || null)
      .attr('x2', (d) => x(d.target.temporal))
      .attr('y2', (d) => y(d.target.bin) || null)
      .merge(links);

    links
      .exit()
      .remove();

    //    Add the nodes to the graph
    const nodes = this.svg
      .selectAll<SVGCircleElement, BaseType>('.node')
      .data(this.data.nodes, (data: IEventElement) => data.id);

    nodes
      .enter()
      .append('circle')
      .attr('class', 'node')
      .attr('cx', (d) => x(d.temporal))
      .attr('cy', (d) => y(d.bin) || null)
      .attr('r', 9)
      .attr('name', (d) => d.bin)
      .attr('node-id', (d) => d.id)
      .attr('entity', (d) => d.entity)
      .style('fill', (d: IEventElement) => z(d.entity))
      .style('opacity', (d) => d.continuing ? 0.7 : 1.0)
      .merge(nodes);

    nodes
      .exit()
      .remove();

    // Add the labels
    const mainLabels = this.svg.selectAll<SVGTextElement, BaseType>('.entityName')
      .data(this.data.nodes, (d: IEventElement) => d.entity);

    mainLabels
      .enter()
      .append('text')
      .text((d) => this.applyFilter(d))
      .attr('class', 'entityName')
      .attr('x', 0)
      .attr('y', (d) => y(d.bin) || 0)
      .attr('text-anchor', 'start')
      .attr('dy', '.1ex')
      .attr('font-size', '12px')
      .merge(mainLabels);

    mainLabels
      .exit()
      .remove();

    //    Update the X-axis
    const xSelection = this.svg.selectAll('g.x-axis');
    if (xSelection.empty()) {
      this.svg
        .append('g')
        .attr('class', 'axis x-axis')
        .attr('transform', 'translate(0,' + this.height + ')')
        .call(axisBottom(x));
    } else {
      xSelection
        .call(axisBottom(x));
    }
  }

  private setupD3(): void {
    this.host = select<HTMLElement, IEventElement | IEventLink>(this.htmlElement);
    console.debug('Width for event graph:', this.htmlElement.offsetWidth);
    this.margin = {top: 20, right: 30, bottom: 20, left: 20};
    // this.width = this.htmlElement.offsetWidth - this.margin.left - this.margin.right;
    // Calculate width
    // If the element isn't actually on the screen, then the width will be 0 which is wrong
    // So we'll either take it from the input, or just make something up
    if (this.graphWidth === undefined) {
      if (this.htmlElement.offsetWidth == 0) {
        this.width = 800 - this.margin.left - this.margin.right;
      } else {
        this.width = (this.htmlElement.offsetWidth) - this.margin.left - this.margin.right;
      }
    } else {
      this.width = this.graphWidth - this.margin.left - this.margin.right;
    }

    this.height = this.graphHeight - this.margin.top - this.margin.bottom;

    // Add the tooltip
    this.tooltip = select('body')
      .append('div')
      // .style("position", "absolute")
      .style('z-index', '10')
      .style('visibility', 'hidden')
      .style('background', 'steelblue')
      .text('a simple tooltip');

    this.svg = this.host.html('')
      .append('svg')
      .attr('width', this.width + this.margin.left + this.margin.right)
      .attr('height', this.height + this.margin.top + this.margin.bottom)
      .append('g')
      .attr('transform', 'translate(' + this.margin.left + ',' + this.margin.top + ')');

    console.debug('Event Graph Initialized');
  }

  /**
   * Apply the filter function, if it exists, otherwise, return the entity as the label
   * @param {IEventElement} input
   * @returns {string}
   */
  private applyFilter = (input: IEventElement): string => {
    if (this.filterLabel) {
      return this.filterLabel(input);
    } else {
      return input.entity;
    }
  };


}
