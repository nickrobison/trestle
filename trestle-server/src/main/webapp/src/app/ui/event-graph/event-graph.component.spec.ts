import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';

import { EventGraphComponent } from './event-graph.component';

describe('EventGraphComponent', () => {
  let component: EventGraphComponent;
  let fixture: ComponentFixture<EventGraphComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ EventGraphComponent ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(EventGraphComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toMatchSnapshot();
  });
});
