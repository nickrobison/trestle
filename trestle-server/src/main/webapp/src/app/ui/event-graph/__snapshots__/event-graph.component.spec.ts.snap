// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventGraphComponent should create 1`] = `
EventGraphComponent {
  "applyFilter": [Function],
  "element": ElementRef {
    "nativeElement": <div>
      <svg
        height="NaN"
        width="800"
      >
        <g
          transform="translate(20,20)"
        />
      </svg>
    </div>,
  },
  "height": NaN,
  "host": Selection {
    "_groups": Array [
      Array [
        <div>
          <svg
            height="NaN"
            width="800"
          >
            <g
              transform="translate(20,20)"
            />
          </svg>
        </div>,
      ],
    ],
    "_parents": Array [
      null,
    ],
  },
  "htmlElement": <div>
    <svg
      height="NaN"
      width="800"
    >
      <g
        transform="translate(20,20)"
      />
    </svg>
  </div>,
  "margin": Object {
    "bottom": 20,
    "left": 20,
    "right": 30,
    "top": 20,
  },
  "svg": Selection {
    "_groups": Array [
      Array [
        <g
          transform="translate(20,20)"
        />,
      ],
    ],
    "_parents": Array [
      null,
    ],
  },
  "tooltip": Selection {
    "_groups": Array [
      Array [
        <div
          style="z-index: 10; visibility: hidden; background: steelblue;"
        >
          a simple tooltip
        </div>,
      ],
    ],
    "_parents": Array [
      <html>
        <head />
        <body>
          <div
            id="root0"
          >
            <div />
            <div>
              <svg
                height="NaN"
                width="800"
              >
                <g
                  transform="translate(20,20)"
                />
              </svg>
            </div>
          </div>
          <div
            style="z-index: 10; visibility: hidden; background: steelblue;"
          >
            a simple tooltip
          </div>
        </body>
      </html>,
    ],
  },
  "tooltipEl": ElementRef {
    "nativeElement": <div />,
  },
  "width": 750,
}
`;
